### PR TITLE
[Core] Implicit resources

### PR DESCRIPTION
**Pull request by jjyao**
## Why are these changes needed?
Add an (experimental) internal concept of implicit resources: resources starts with a certain prefix are implicit resources and all nodes implicitly have them with total capacity of 1.

Several key implementation decisions:

1. The code that deals with implicit resources is limited to `NodeResourceSet` and `NodeResourceInstanceSet`.
2. For existing Ray applications that don't use implicit resources, the behavior is unchanged.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
Note: See https://docs.google.com/document/d/151T4VnknX_5EtPy6E-LbpL-r1T4ZSO0UBvSgWdSjx4Q/edit#heading=h.2ews5m5fmz for more context. 

The decision is that it is the private API, and changing implementation in the future is not very difficult. We need more refinement in observabiltiy (@jjyao will follow up). In the long-term, we should have general way to handle affinity/anti-affinity (this is a special case anti-affinity).

File: python/ray/_private/ray_constants.py [hunk 1]
```diff
@@ -437,6 +437,9 @@ def gcs_actor_scheduling_enabled():
 
 DEFAULT_RESOURCES = {"CPU", "GPU", "memory", "object_store_memory"}
 
+# Sync with src/ray/common/constants.h
```
Comment (rkooo567) [inline python/ray/_private/ray_constants.py:440]: I remember it was possible to share the same variable between this & constant.h? Can we just do it?

File: python/ray/autoscaler/_private/resource_demand_scheduler.py [hunk 2]
```diff
@@ -932,7 +933,9 @@ def get_bin_pack_residual(
 
 def _fits(node: ResourceDict, resources: ResourceDict) -> bool:
     for k, v in resources.items():
-        if v > node.get(k, 0.0):
+        if v > node.get(
```
Comment (rkooo567) [inline python/ray/autoscaler/_private/resource_demand_scheduler.py:936]: Do we have unit tests to cover implicit resources autoscaling? 

- Test we have max_relica_per_node = 1 & schedule 5 actors. Make sure 5 instances are started at the same time.

File: python/ray/autoscaler/_private/resource_demand_scheduler.py [hunk 3]
```diff
@@ -932,7 +933,9 @@ def get_bin_pack_residual(
...
+            k, 1.0 if k.startswith(ray_constants.IMPLICIT_RESOURCE_PREFIX) else 0.0
```
Comment (rkooo567) [inline python/ray/autoscaler/_private/resource_demand_scheduler.py:937]: Q: should we just hide this logic under ResourceDict instead of exposing it here?

Comment (rkooo567) [reply python/ray/autoscaler/_private/resource_demand_scheduler.py:937]: Same for the below logic

File: src/ray/common/scheduling/resource_set.h [hunk 4]
```diff
@@ -40,7 +39,7 @@ class ResourceSet {
   }
 
   /// \brief Empty ResourceSet constructor.
-  ResourceSet();
+  ResourceSet(){};
```
Comment (rkooo567) [inline src/ray/common/scheduling/resource_set.h:42]:
```suggestion
  ResourceSet() {};
```

Comment (jjyao) [reply src/ray/common/scheduling/resource_set.h:42]: I think clang-format doesn't agree :)

**Review by rkooo567 — APPROVED** [review]
There's a concern that ResourceSet and NodeResourceSet relationship is confusing, but we will rename ResourceSet to immutable ResourceDemand class as a follow up, then it LGTM

File: python/ray/autoscaler/_private/resource_demand_scheduler.py (hunk 3 above)
Comment (jjyao) [reply python/ray/autoscaler/_private/resource_demand_scheduler.py]: Currently ResourceDict is just an alias of python dict. Added a TODO to wrap it in a class.

**Comment by Irvingwangjr** [comment]
> Note: See https://docs.google.com/document/d/151T4VnknX_5EtPy6E-LbpL-r1T4ZSO0UBvSgWdSjx4Q/edit#heading=h.2ews5m5fmz for more context.
> 
> The decision is that it is the private API, and changing implementation in the future is not very difficult. We need more refinement in observabiltiy (@jjyao will follow up). In the long-term, we should have general way to handle affinity/anti-affinity (this is a special case anti-affinity).

can you make this document public available?

**Pull request merged**